pkg/openstack/client/client.go

```diff
@@ -0,0 +1,140 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package client
+
+import (
+	"context"
+
+	os "github.com/gardener/gardener-extension-provider-openstack/pkg/openstack"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/utils/openstack/clientconfig"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func NewOpenstackClientFromCredentials(credentials *os.Credentials) (Factory, error) {
+	opts := &clientconfig.ClientOpts{
+		AuthInfo: &clientconfig.AuthInfo{
+			AuthURL:     credentials.AuthURL,
+			Username:    credentials.Username,
+			Password:    credentials.Password,
+			ProjectName: credentials.TenantName,
+			DomainName:  credentials.DomainName,
+		},
+	}
+
+	authOpts, err := clientconfig.AuthOptions(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	// AllowReauth should be set to true if you grant permission for Gophercloud to
+	// cache your credentials in memory, and to allow Gophercloud to attempt to
+	// re-authenticate automatically if/when your token expires.
+	authOpts.AllowReauth = true
+
+	provider, err := openstack.AuthenticatedClient(*authOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &OpenstackClientFactory{
+		providerClient: provider,
+	}, nil
+}
+
+func NewOpenstackClientFromSecretRef(ctx context.Context, c client.Client, secretRef corev1.SecretReference) (Factory, error) {
+	creds, err := os.GetCredentials(ctx, c, secretRef)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewOpenstackClientFromCredentials(creds)
+}
+
+func Region(region string) Option {
+	return func(opts gophercloud.EndpointOpts) gophercloud.EndpointOpts {
+		opts.Region = region
+		return opts
+	}
+}
+
+func (oc *OpenstackClientFactory) Storage(options ...Option) (Storage, error) {
+	eo := gophercloud.EndpointOpts{}
+	for _, opt := range options {
+		eo = opt(eo)
+	}
+	storageClient, err := openstack.NewObjectStorageV1(oc.providerClient, gophercloud.EndpointOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	return &StorageClient{
+		client: storageClient,
+	}, nil
+}
+
+type serviceVersion struct {
+	Version struct {
+		Version    string `json:"version"`
+		Status     string `json:"status"`
+		MinVersion string `json:"min_version"`
+	} `json:"version"`
```

Comment:
Why are the fields exported?
Why is this an inline struct leading to json paths like `version.version`?

Reply:
a) The fields are exported because the AFAIK the JSON decode only takes into account (thus filling) exported fields.
b) I was confused too but that's what openstack returns for the call. The struct is a 1-to-1 match. [Source](1), but the call we use does not return a list and the return json looks like: 
```json
{
  "version": {
      {
          "id": "v2.0",
          "links": [
              {
                  "href": "http://openstack.example.com/v2/",
                  "rel": "self"
              }
          ],
          "status": "SUPPORTED",
          "version": "",
          "min_version": "",
          "updated": "2011-01-21T11:33:21Z"
      }
}
```
As for inlining the struct, its used and will be used only in this place to read the response. My reasoning was to not introduce symbols with practically 0 reusability. 

The naming of the fields is to avoid `VersionOuter` variants and create more confusion.

[1]: https://docs.openstack.org/api-guide/compute/microversions.html

Reply:
Interesting, thanks for clarifying!